datafusion-examples/examples/parquet_encryption_with_kms.rs

```diff
@@ -0,0 +1,205 @@
+/// parquet-key-management crate to integrate with a Key Management Server (KMS).
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let ctx = SessionContext::new();
+
+    // Register an `EncryptionFactory` implementation to be used for Parquet encryption
+    // in the session context.
+    // This uses an in-memory test KMS from the `parquet_key_management` crate with
+    // a custom `KmsEncryptionFactory` wrapper type to integrate with DataFusion.
+    // `EncryptionFactory` instances are registered with a name to identify them so
+    // they can be later referenced in configuration options, and it's possible to register
+    // multiple different factories to handle different ways of encrypting Parquet.
+    // In future it could be possible to have built-in implementations in DataFusion.
+    let crypto_factory = CryptoFactory::new(TestKmsClientFactory::with_default_keys());
+    let encryption_factory = KmsEncryptionFactory { crypto_factory };
+    ctx.register_parquet_encryption_factory(
+        "example.inmem_kms_encryption",
+        Arc::new(encryption_factory),
+    );
+
+    // Register options required for our encryption factory.
+    // This works similarly to register_table_options_extension but registers extension options
+    // specific to Parquet, so they can be set in `TableParquetOptions`.
+    ctx.register_parquet_options_extension(EncryptionConfig::default());
+
+    let tmpdir = TempDir::new()?;
+    write_encrypted(&ctx, &tmpdir).await?;
+    read_encrypted(&ctx, &tmpdir).await?;
+    Ok(())
+}
+
+/// Write an encrypted Parquet file
+async fn write_encrypted(ctx: &SessionContext, tmpdir: &TempDir) -> Result<()> {
+    let a: ArrayRef = Arc::new(StringArray::from(vec!["a", "b", "c", "d"]));
+    let b: ArrayRef = Arc::new(Int32Array::from(vec![1, 10, 10, 100]));
+    let batch = RecordBatch::try_from_iter(vec![("a", a), ("b", b)])?;
+
+    ctx.register_batch("test_data", batch)?;
+    let df = ctx.table("test_data").await?;
+
+    let mut parquet_options = TableParquetOptions::new();
+    // We specify that we want to use Parquet encryption by setting the name of the
+    // encryption factory to use.
+    parquet_options.global.encryption_factory = "example.inmem_kms_encryption".to_owned();
+    // Our encryption factory requires specifying the master key identifier to
+    // use for encryption:
+    let encryption_options = parquet_options
+        .extensions
+        .get_mut::<EncryptionConfig>()
+        .ok_or_else(|| {
+            DataFusionError::Configuration(
+                "EncryptionConfig options not registered".to_owned(),
+            )
+        });
+    encryption_options.footer_key_id = Some("kf1".to_owned());
+    // Question: I don't think extensions.get will work when creating new `TableParquetOptions`,
+    // we probably need to register the extension options here but also allow registering them
+    // on the session context for compatibility with setting options via strings?.
```

Comment:
As mentioned, I think it would be better if we add these structures directly to `struct ParquetOptions`. In what I did I build a string version that could hold the parameters with `from` and `into` support for the underlying information.